```{r setup, include=FALSE}
library(data.table)
library(seqinr)
library(stringr)
library(stats)
library(dplyr)
library(RGenetics)
library(splus2R)
library(magrittr)
library(rlist)
library(ape)
library(ggplot2)
library(gridExtra)
library(ggpubr)
library(ggExtra)
library(corrplot)
library(viridis)
library(wesanderson)
library(latex2exp)
library(grid)
library(gridExtra)
library(cowplot)
library(readr)
library(ggtree)
library(factoextra)
library(directlabels)
library(ggfortify)
library(devtools)
library(ggbiplot)
library(FactoMineR)
library(ggpubr)
library(tidyverse)
library(ggrepel)
library(ggbeeswarm)
library(grImport)
```

Create fasta file with alignment with XXXX. Even coordinates for all vvariants.
ORF1ab = leader (180 ac) + nsp2 (638 ac) + nsp3 (1945 ac) + nsp4 (500 ac) + 3C (306 ac) + nsp6 (290 ac) + nsp7 (83 ac) +
nsp8 (198 ac) + nsp9 (113 ac) + nsp10(139 ac) + RdRp (932 ac) + helicase (601 ac) + exo (527 ac) + endo (346 ac) + methyl (298 ac) + 
nsp11 (13 ac)

######################################################################
######################################################################
######################################################################
General for HLA I and HLA II
mutations table

```{r setup, include=FALSE}
path <- "/home/jane/PhD_Skoltech/coronavirus/immunocompromised/"
ref <- as.character(read.FASTA(paste0(path, "russian.fa"), type = "DNA"))

coord <- rbind(c("leader", 266-100, 805-100),
               c("nsp2", 806-100, 2719-100),
               c("nsp3",  2720-100, 8554-100),
               c("nsp4", 8555-100, 10054-100),
               c("3C", 10055-100, 10972-100),
               c("nsp6", 10973-100, 11842-100),
               c("nsp7", 11843-100, 12091-100),
               c("nsp8", 12092-100, 12685-100),
               c("nsp9", 12686-100, 13024-100),
               c("nsp10",  13025-100, 13441-100),
               c("nsp11", 13442-100, 13480-100),
               c("RdRp", "13342:13368", "13368:16136"),
               c("helicase", 16237-100, 18039-100),
               c("exonuclease", 18040-100, 19620-100),
               c("endornase",  19621-100, 20658-100),
               c("methyltransferase", 20659-100, 21552-100),
               c("S", 21563-100, 25384-100),
               c("ORF3a", 25393-100, 26220-100),
               c("E", 26245-100, 26472-100),
               c("M", 26523-100, 27191-100), 
               c("ORF6", 27202-100, 27387-100),
               c("ORF7a", 27394-100, 27759-100),
               c("ORF7b", 27756-100, 27887-100),
               c("ORF8", 27894-100, 28259-100),
               c("N", 28274-100, 29533-100),
               c("ORF10", 29558-100, 29674-100))

coord <- data.frame(coord)
colnames(coord) <- c("gene", "start", "end")

seq <- c()
for (g in coord$gene){
  if (g != "RdRp"){
    coordinates <-  as.numeric(as.character(coord$start[coord$gene == g])):as.numeric(as.character(coord$end[coord$gene == g]))
  } else {
    coordinates <- c(as.numeric(unlist(str_split(coord$start[coord$gene == g], ":"))[1]):
                       as.numeric(unlist(str_split(coord$start[coord$gene == g], ":"))[2]),
                     as.numeric(unlist(str_split(coord$end[coord$gene == g], ":"))[1]):
                       as.numeric(unlist(str_split(coord$end[coord$gene == g], ":"))[2]))
  }
  
  seq_g <- translate(ref[[1]][coordinates])
  if (g == "ORF8"){
    seq_g[18] <- "Q"
  }
  seq_g <- paste0(seq_g, collapse = "")
  if (substring(seq_g, nchar(seq_g), nchar(seq_g)) == "*") {seq_g = substring(seq_g, 1, nchar(seq_g)-1)}
  seq <- c(seq, seq_g)
  
}
coord$seq <- seq
#coord[, c(2, 3)] <- sapply(coord[, c(2, 3)], as.numeric)


mutations <- rbind(c("C:21711:T", 50, 50, "S", "L", "S", "subst", "S:S50L", "Aug", "S", "AJ", 0.999700142857143, "+"),
                   c("ATACATG:21764:A", 69, 70, "IHV", "XXI", "S", "del", "S:del68_70", "Aug", "S", "A", 0.216505428571429, "+"),
                   ## name!
                   c("TTTTGGGTGTTTA:21981:T", 140, 144, "FLGVY", "XXXXX", "S", "del", "S:del140_144", "Aug", "S", "AJ", 0.968835, "+"),
                   c("G:22381:T", 273, 273, "R", "S", "S", "subst", "S:R273S", "Jan", "S", "J", 0.286, "+"),
                   c("T:22882:G", 440, 440, "N", "K", "S", "subst", "S:N440K", "Jan", "S", "J", 0.428, "+"),
                   c("T:22917:G", 452, 452, "L", "R", "S", "subst", "S:L452R", "Jan", "S", "J", 0.428, "-"),
                   c("A:22920:T", 453, 453, "Y", "F", "S", "subst", "S:Y453F", "Aug", "S", "A", 0.313, "+"),
                   c("C:22971:A", 470, 470, "T", "N", "S", "subst", "S:T470N", "Aug", "S", "AJ", 0.713, "+"),
                   c("G:22988:A", 476, 476, "G", "S", "S", "subst", "S:G476S", "Jan", "S", "J", 0.571, "-"),
                   c("C:23423:T", 621, 621, "P", "S", "S", "subst", "S:P621S", "Aug", "S", "A", 0.149, "-"),
                   c("A:23772:G", 737, 737, "D", "G", "S", "subst", "S:D737G", "Aug", "S", "AJ", 0.996, "-"),
                   c("C:2037:T", 411, 411, "A", "V", "nsp2", "subst", "nsp2:A411V", "Jan", "S", "J", 0.571, "+"),
                   c("A:4229:G", 504, 504, "T", "A", "nsp3", "subst", "nsp3:T504A", "Aug", "S", "A", 0.239, "+"),
                   c("A:4229:C", 504, 504, "T", "P", "nsp3", "subst", "nsp3:T504P", "Jan", "S", "J", 0.570, "+"),
                   c("G:5180:A", 821, 821, "D", "N", "nsp3", "subst", "nsp3:D821N", "Aug", "S", "AJ", 0.857, "+"),
                   c("C:7086:T", 1456, 1456, "T", "I", "nsp3", "subst", "nsp3:T1456I", "Jan", "S", "J", 0.428, "+"),
                   c("C:9438:T", 295, 295, "T", "I", "nsp4", "subst", "nsp4:T295I", "Aug", "S", "AJ", 0.999, "-"),
                   c("G:9497:A", 315, 315, "V", "I", "nsp4", "subst", "nsp4:V315I", "Aug", "S", "AJ", 0.813, "-"),
                   c("TG:11082:T", 37, 37, "T", "X", "nsp6", "del", "nsp6:del37_37", "Aug", "S", "AJ", 0.135789142857143, "-"),
                   c("G:11083:T", 37, 37, "L", "F", "nsp6", "subst", "nsp6:L37F", "Jan", "S", "AJ", 0.413626, "-"),
                   ##
                   c("", 183, 183, "M", "I", "nsp6", "subst", "nsp6:M183I", "Aug", "S", "A", 2, "-"),
                   c("G:11804:A", 278, 278, "V", "I", "nsp6", "subst", "nsp6:V278I", "Jan", "S", "J", 0.068, "+"),
                   c("T:12015:G", 58, 58, "V", "G", "nsp7", "subst", "nsp7:V58G", "Aug", "S", "AJ", 0.998, "+"),
                   c("A:13913:G", 158, 158, "N", "S", "RdRp", "subst", "RdRp:N158S", "Jan", "S", "J", 0.428, "-"),
                   ##
                   c("", 263, 263, "V", "I", "exonuclease", "subst", "exonuclease:V263I", "Jan", "S", "A", 2, "-"),
                   c("C:20234:T", 205, 205, "P", "L", "endornase", "subst", "endornase:P205L", "Jan", "S", "J",
                     0.571, "+"),
                   ##
                   c("", 94, 94, "L", "F", "ORF3a", "subst", "ORF3a:L94F", "Aug", "S", "A", 2, "-"),
                   ## ORF&a not ORF7b???
                   c("GATT:27758:G", 2, 2, "Q", "X", "ORF7a", "del", "ORF7a:del2_2", "Aug", "S", "AJ", 0.855, "-"),
                    ## ORF&a not ORF7b???
                   c("GATT:27758:G", 2, 2, "Q", "X", "ORF7b", "del", "ORF7b:del2_2", "Aug", "S", "AJ", 2, "-"),
                   c("C:27945:T", 18, 18, "Q", "*", "ORF8", "stop", "ORF8:Q18*", "Aug", "S", "AJ", 0.854, "+"),
                   c("C:28289:A", 6, 6, "P", "T", "N", "subst", "N:P6T", "Aug", "S", "AJ", 0.601, "+"),
                   c("A:28856:G", 195, 195, "R" , "G", "N", "subst", "N:R195G", "Aug", "S", "A", 0.231, "-"),
                   c("C:26261:T", 6, 6, "S", "L", "E", "subst", "E:S6L", "Aug", "S", "AJ", 0.580, "-"), 
                   c("T:26320:C", 26, 26, "F", "L", "E", "subst", "E:F26L", "Jan", "S", "J", 0.286, "-"),
                   c("T:26908:G", 129, 129, "L", "R", "M", "subst", "M:L129R", "Aug", "S", "AJ", 0.464, "+"))
mutations <- data.frame(mutations)
colnames(mutations) <- c("g_coord", "pos", "end", "or", "mut", "gene", "mut_type", "name", "date", "patient", "dyn", "freq", "immuno")
mutations[, c(2, 3, 12)] <- sapply(mutations[, c(2, 3, 12)], as.numeric)
mutations$g_pos <- as.numeric(str_extract(mutations$g_coord, "[0-9][0-9]*"))
mutations$g_or <- str_extract(mutations$g_coord, "[A-Z][A-Z]*:")
mutations$g_or <- substring(mutations$g_or, 1, nchar(mutations$g_or)-1)
mutations$g_mut <- substring(str_extract(mutations$g_coord, ":[A-Z][A-Z]*"), 2)
write.csv(mutations, "/home/jane/PhD_Skoltech/coronavirus/hla_pop_mut/patient_mutations.csv", row.names = FALSE)
```


Writing peptides 
8-14 for mhc I
12-18 for mhc II
19 - for the case if there is X in addition to others mut-s


```{r setup, include=FALSE}
peptides_table <- c()
for (m in 1:nrow(mutations)){
  ## general stuff  
  seq_or <- coord$seq[coord$gene == mutations$gene[m]]
  substring(seq_or, mutations$pos[m], mutations$end[m]) <- mutations$or[m]
  seq_mut <- seq_or
  substring(seq_mut, mutations$pos[m], mutations$end[m]) <- mutations$mut[m]
    
  ## for substitution mutations
  if (mutations$mut_type[m] == "subst") {
    for (i in 8:19){
      mut_start <- mutations$pos[m] - i + 1
      mut_end <- mutations$end[m] + i - 1
      if (mut_start <= 0) {mut_start = 1}
      if (mut_end >= nchar(seq_or)) {mut_end >= nchar(seq_or)}
      mut_region_or <- substring(seq_or, mut_start, mut_end)
      mut_region_mut <- substring(seq_mut, mut_start, mut_end)
        
      for (j in 1:(nchar(mut_region_or) - i + 1)) {
        pept_or <- substring(mut_region_or, j, j + i - 1)
        pept_mut <- substring(mut_region_mut, j, j + i - 1)
        pept_or <- str_remove_all(pept_or, "X")
        pept_mut <- str_remove_all(pept_mut, "X")
        if (nchar(pept_or) >= 8 & nchar(pept_or) <= 18){
        peptides_table <- rbind(peptides_table, 
                          c(mutations$gene[m], mutations$mut_type[m], mutations$name[m], pept_or, pept_mut, nchar(pept_or)))
        }
      }
    }
  
  ## for deletions
  ## by genome coordinates
  } else if (mutations$mut_type[m] == "del"){
    
    gene_or <- toupper(paste0(ref[[1]][as.numeric(as.character(coord$start[coord$gene == mutations$gene[m]])):
                           as.numeric(as.character(coord$end[coord$gene == mutations$gene[m]]))], collapse = "")) 
    substring(gene_or, mutations$g_pos[m] - 100 - as.numeric(coord$start[coord$gene == mutations$gene[m]]) + 1, 
            mutations$g_pos[m] - 100 - as.numeric(coord$start[coord$gene == mutations$gene[m]]) + 1 + 
              nchar(mutations$g_or[m]) - 1) <- mutations$g_or[m]   
     
    gene_mut <- gene_or
    substring(gene_mut, mutations$g_pos[m] - 100 - as.numeric(coord$start[coord$gene == mutations$gene[m]]) + 1, 
            mutations$g_pos[m] - 100 - as.numeric(coord$start[coord$gene == mutations$gene[m]]) + 1 + 
              nchar(mutations$g_or[m]) - 1) <- paste0(mutations$g_mut[m], rep("-",  nchar(mutations$g_or[m]) - 1), 
                                                      collapse = "")   
     
    pr_or <- paste0(translate(unlist(str_split(gene_or, ""))), collapse = "")
    pr_mut <- paste0(translate(unlist(str_split(gene_mut, ""))), collapse = "")
      
    del_size <- nchar(mutations$or[m])
    for (i in 8:19){
      mut_start <- mutations$pos[m] - i + 1
      mut_end <- mutations$end[m] + i - 1
      if (mut_start <= 0) {mut_start = 1}
      if (mut_end >= nchar(pr_or)) {mut_end >= nchar(pr_or)}
      mut_region_or <- substring(pr_or, mut_start, mut_end)
      mut_region_mut <- substring(pr_mut, mut_start, mut_end)
      mut_region_mut <- paste0(str_remove_all(mut_region_mut, "X"), 
                               substring(pr_mut, mut_end + 1, mut_end + str_count(mut_region_mut, "X")), collapse = "")
        
      for (j in 1:(nchar(mut_region_or) - i + 1)) {
        pept_or <- substring(mut_region_or, j, j + i - 1)
        pept_mut <- substring(mut_region_mut, j, j + i - 1)
        ## if we can prolongate mut peptide because of deletion
        ## the end of peptide is less then seq
        #if(mut_start + j + i - 2 < nchar(seq_mut)){
         #  after_peptide <- substring(seq_mut, mut_start +j + i - 1, nchar(seq_mut))
          # if (del_size <= after_peptide){del_add <- substring(after_peptide, 1, del_size)} else {del_add <- after_peptide}
           #pept_mut <- str_remove_all(paste0(pept_mut, del_add, collapse = ""), "X")
           ### if not all deletion included
           #pept_mut <- substring(pept_mut, 1, nchar(pept_or))
        #} else {pept_mut = "-"}
         
        pept_or <- str_remove_all(pept_or, "X")
        
        if (nchar(pept_or) >= 8 & nchar(pept_or) <= 18){
        peptides_table <- rbind(peptides_table, 
                           c(mutations$gene[m], mutations$mut_type[m], mutations$name[m], pept_or, pept_mut, nchar(pept_or)))
        }
      }
    }
    
  ## for stops
  } else if (mutations$mut_type[m] == "stop"){
    for (i in 8:19){
      mut_start <- mutations$pos[m] - i + 1
      mut_end <- nchar(seq_or)
      if (mut_start <= 0) {mut_start = 1}
      mut_region_or <- substring(coord$seq[coord$gene == mutations$gene[m]], mut_start, mut_end)
      
      for (j in 1:(nchar(mut_region_or) - i + 1)) {
        pept_or <- substring(mut_region_or, j, j + i - 1)
        pept_mut <- "-"
        pept_or <- str_remove_all(pept_or, "X")
        if (nchar(pept_or) >= 8 & nchar(pept_or) <= 18){
          peptides_table <- rbind(peptides_table, 
                            c(mutations$gene[m], mutations$mut_type[m], mutations$name[m], pept_or, pept_mut, nchar(pept_or)))
        }
      } 
    }
  }
}  
peptides_table <- data.frame(peptides_table)
colnames(peptides_table) <- c("gene", "mut_type", "mut", "pept_or", "pept_mut", "l")
peptides_table[, c(1:6)] <- sapply(peptides_table[, c(1:6)], as.character)
peptides_table[, 6] <- as.numeric(peptides_table[, 6])
peptides_table <-  distinct(peptides_table, .keep_all = FALSE)
saveRDS(peptides_table, file = paste0(path, "peptides_table_8_18.Rds", sep = ""))

peptides_I <- c(peptides_table$pept_or[peptides_table$l >= 8 & peptides_table$l <= 14], 
              peptides_table$pept_mut[peptides_table$l >= 8 & peptides_table$l <= 14 & peptides_table$pept_mut != "-"])

write.table(peptides_I, quote = FALSE, sep = "\n", col.names = FALSE, row.names = FALSE, file = paste0(path, "mhcI.pept", sep = ""))

peptides_II <- c(peptides_table$pept_or[peptides_table$l >= 12 & peptides_table$l <= 18], 
              peptides_table$pept_mut[peptides_table$l >= 12 & peptides_table$l <= 18 & peptides_table$pept_mut != "-"])

write.table(peptides_II, quote = FALSE, sep = "\n", col.names = FALSE, row.names = FALSE, file = paste0(path, "mhcII.pept", sep = ""))

```


###################################################
###################################################
Count peptides in ranks ....
net_pan for hla I
calculate best rank

```{r}
path = "/home/jane/PhD_Skoltech/coronavirus/immunocompromised/net_pan/"
peptides_table <- readRDS("/home/jane/PhD_Skoltech/coronavirus/immunocompromised/peptides_table_8_18.Rds")

net_pan_files <- list.files(path, pattern = "*filtered")
########
########
## mhc I or II
net_pan_files <- net_pan_files[!str_detect(net_pan_files, "mhcII")]
net_panI <- c()

for (f in net_pan_files){
  net_pan_f <- read_tsv(paste0(path, f, collapse = ""), col_names = FALSE) %>%  separate(X1, c("N", "allele", "peptide", "core", "a", "b", "c", "d", "e", "Icore", "gene_short", "Score_EL", "Rank_EL", "Score_BA", "Rank_BA", "Aff", "type"), sep = "\\s+")
  net_pan_f <- data.frame(net_pan_f[,c(1:17)])
  net_pan_f[, c(1,5:9,12:16)] <- sapply(net_pan_f[, c(1,5:9,12:16)], as.numeric)
  net_pan_f <- unique(net_pan_f[,c(2,3,4,10,12:17)])
  #net_pan_f <- net_pan_f[2:nrow(net_pan_f),]
  net_panI <- rbind(net_panI, net_pan_f)
}

  
add <- c()
alleles <- unique(net_panI$allele)
#alleles <- alleles[alleles != ":"]

# Rank_EL thresholds
thresholds <- c(0, 0.5, 2)


########
########
## mhc I or II
peptides <- peptides_table[nchar(peptides_table$pept_or) <= 11,]
peptides$mut[peptides$mut == "S:del69-70"] <- "S:del68_70"
peptides$mut[peptides$mut == "S:del141-144"] <- "S:del140_144"
peptides$mut[peptides$mut == "nsp6:del37"] <- "nsp6:del37_37"
peptides$mut[peptides$mut == "ORF7a:del2"] <- "ORF7a:del2_2"

best_rank_alleles <- c()
muts <- unique(peptides$mut)
muts <- muts[!is.element(muts, c("ORF7b:del2_2", "nsp6:M183I", "exonuclease:V263I", "ORF3a:L94F"))] 
#muts <- muts[muts!="ORF8:Q18*"]
for (i in muts){
  peptides_or <- peptides$pept_or[peptides$mut == i]
  peptides_mut <- peptides$pept_mut[peptides$mut == i]
  rank_or <- c()
  rank_mut <- c()
  for (al in alleles){
     rank_or <- c(rank_or, min(net_panI$Rank_EL[is.element(net_panI$peptide, peptides_or) & net_panI$allele == al]))
     if (i == "ORF8:Q18*"){
       rank_mut <- c(rank_mut, 100)
     } else {
      rank_mut <- c(rank_mut, min(net_panI$Rank_EL[is.element(net_panI$peptide, peptides_mut) & net_panI$allele == al]))
     }
  }
  best_rank_alleles <- rbind(best_rank_alleles, 
                                c(i, "or", rank_or))
  best_rank_alleles <- rbind(best_rank_alleles, 
                                c(i, "mut", rank_mut))
  best_rank_alleles <- rbind(best_rank_alleles, 
                                c(i, "delta", rank_or - rank_mut))
  best_rank_alleles <- rbind(best_rank_alleles, 
                                c(i, "dev", rank_mut/rank_or))
}
 
best_rank_alleles <- data.frame(best_rank_alleles)    
colnames(best_rank_alleles) <- c("mut", "state", alleles)
best_rank_alleles[, 3:97] <- sapply(best_rank_alleles[, 3:97], as.numeric)
saveRDS(best_rank_alleles, paste0(path, "best_rank_mhcI.Rds", collapse = ""))
```

net_pan for hla II
calculate best rank

```{r}
path = "/home/jane/PhD_Skoltech/coronavirus/immunocompromised/net_pan/"
peptides_table <- readRDS("/home/jane/PhD_Skoltech/coronavirus/immunocompromised/peptides_table_8_18.Rds")

net_pan_files <- list.files(path, pattern = "*filtered")
########
########
## mhc I or II
net_pan_files <- net_pan_files[str_detect(net_pan_files, "mhcII")]
net_pan <- c()

for (f in net_pan_files){
  net_pan_f <- read_tsv(paste0(path, f, collapse = ""), col_names = FALSE) %>%  separate(X1, c("N", "allele", "peptide", "core", "a", "b", "c", "d", "e", "Icore", "gene_short", "Score_EL", "Rank_EL", "Score_BA", "Rank_BA", "Aff", "type"), sep = "\\s+")
  net_pan_f <- data.frame(net_pan_f[,c(1:17)])
  net_pan_f[, c(1,5:9,12:16)] <- sapply(net_pan_f[, c(1,5:9,12:16)], as.numeric)
  net_pan_f <- unique(net_pan_f[,c(2,3,4,10,12:17)])
  #net_pan_f <- net_pan_f[2:nrow(net_pan_f),]
  net_pan <- rbind(net_pan, net_pan_f)
}

  
add <- c()
alleles <- unique(net_pan$allele)
#alleles <- alleles[alleles != ":"]

# Rank_EL thresholds
thresholds <- c(0, 0.5, 2)


########
########
## mhc I or II
peptides <- peptides_table[nchar(peptides_table$pept_or) >= 12,]
peptides$mut[peptides$mut == "S:del69-70"] <- "S:del68_70"
peptides$mut[peptides$mut == "S:del141-144"] <- "S:del140_144"
peptides$mut[peptides$mut == "nsp6:del37"] <- "nsp6:del37_37"
peptides$mut[peptides$mut == "ORF7a:del2"] <- "ORF7a:del2_2"

best_rank_alleles <- c()
muts <- unique(peptides$mut)
muts <- muts[!is.element(muts, c("ORF7b:del2_2", "nsp6:M183I", "exonuclease:V263I", "ORF3a:L94F"))] 
#muts <- muts[muts!="ORF8:Q18*"]
for (i in muts){
  peptides_or <- peptides$pept_or[peptides$mut == i]
  peptides_mut <- peptides$pept_mut[peptides$mut == i]
  rank_or <- c()
  rank_mut <- c()
  for (al in alleles){
     rank_or <- c(rank_or, min(net_pan$Rank_EL[is.element(net_pan$peptide, peptides_or) & net_pan$allele == al]))
     if (i == "ORF8:Q18*"){
        rank_mut <- c(rank_mut, 100)
     } else {
        rank_mut <- c(rank_mut, min(net_pan$Rank_EL[is.element(net_pan$peptide, peptides_mut) & net_pan$allele == al])) 
     }
  }
  best_rank_alleles <- rbind(best_rank_alleles, 
                                c(i, "or", rank_or))
  best_rank_alleles <- rbind(best_rank_alleles, 
                                c(i, "mut", rank_mut))
  best_rank_alleles <- rbind(best_rank_alleles, 
                                c(i, "delta", rank_or - rank_mut))
  best_rank_alleles <- rbind(best_rank_alleles, 
                                c(i, "dev", rank_mut/rank_or))
}
 
best_rank_alleles <- data.frame(best_rank_alleles)    
colnames(best_rank_alleles) <- c("mut", "state", alleles)
best_rank_alleles[, 3:73] <- sapply(best_rank_alleles[, 3:73], as.numeric)
saveRDS(best_rank_alleles, paste0(path, "best_rank_mhcII.Rds", collapse = ""))
```

##############
PHBR for mutation

```{r}
muts <- unique(mutations$name)
muts <- muts[!is.element(muts, c("ORF7b:del2_2", "nsp6:M183I", "exonuclease:V263I", "ORF3a:L94F"))] 
#muts <- muts[muts!="ORF8:Q18*"]

path = "/home/jane/PhD_Skoltech/coronavirus/immunocompromised/net_pan/"
best_rank_alleles <- readRDS(paste0(path, "best_rank_mhcI.Rds", collapse = ""))
best_rank_alleles[best_rank_alleles$state != "delta", ] %>% 
  melt -> df
s_alleles <- c("HLA-A*01:01", "HLA-A*03:01", "HLA-B*07:02", "HLA-B*08:01", "HLA-C*07:01", "HLA-C*07:02")
s_best <- df[is.element(df$variable, s_alleles), ]

phbrI <- c()
for (i in muts){
  values_or <- s_best$value[s_best$mut == i & s_best$state == "or"]
  values_mut <- s_best$value[s_best$mut == i & s_best$state == "mut"]
  if (sum(values_or <= 2) > 0 | sum(values_mut <= 2) > 0) {
    phbrI <- rbind(phbrI, c(i, 6/sum(1/values_or), 6/sum(1/values_mut), mutations$freq[mutations$name == i],
                          mutations$dyn[mutations$name == i], mutations$immuno[mutations$name == i], "yes"))
  } else {
    phbrI <- rbind(phbrI, c(i, 6/sum(1/values_or), 6/sum(1/values_mut), mutations$freq[mutations$name == i],
                          mutations$dyn[mutations$name == i], mutations$immuno[mutations$name == i], "no"))
  }
  #phbr <- rbind(phbr, c(i, 6/sum(1/values_mut), "mut"))
}
phbrI <- data.frame(phbrI)
colnames(phbrI) <- c("mut", "phbr_or", "phbr_mut", "freq", "dyn", "immuno", "presented")
phbrI[,2:4] <- sapply(phbrI[,2:4], as.numeric)
phbrI$rel <- phbrI$phbr_mut/phbrI$phbr_or

#pal <- wes_palette("Zissou1", 100, type = "continuous")

phbrI$phbr_mut[phbrI$mut == "ORF8:Q18*"] <- 10
phbrI$rel[phbrI$mut == "ORF8:Q18*"] <- 14

phbrI$mut1 <- phbrI$mut
phbrI$mut1[phbrI$rel <= 2] <- NA

#phbrI$rel[phbrI$rel < 1] <- 1/phbrI$rel[phbrI$rel < 1]
p_mutI <- ggplot(phbrI[phbrI$presented == "yes" & phbrI$mut != "ORF8:Q18*",], aes(x = phbr_or, y = phbr_mut, label = mut1)) +
  geom_abline(slope = 1, linetype="dashed", color = "grey10") +
  geom_point(aes(fill = rel), size = 6, shape = 21, color = "grey20", alpha = 0.9) +
  scale_x_log10(limits = c(0.005, 10), name = "PHBR in ancestral state") +
  scale_y_log10(limits = c(0.005, 10), name = "PHBR in derived state") +
  #scale_fill_gradientn(colours = pal) + 
  scale_fill_distiller(name = "PHBR\nFold\nChange", palette = "RdYlBu", limits = c(0.19, 14)) +
  geom_text_repel(size = 6) +
  #ggtitle("HLA I") +
  theme_bw() +
   theme(legend.position = "none",
        axis.text.y = element_text(size = 18),
        axis.text.x = element_text(size = 18),
        legend.title = element_text (size = 18),
        legend.text = element_text(size = 18),
        axis.title = element_text (size = 18))


best_rank_alleles <- readRDS(paste0(path, "best_rank_mhcII.Rds", collapse = ""))
best_rank_alleles[best_rank_alleles$state != "delta", ] %>% 
  melt -> df
s_alleles <- c("DRB1_0101",  "DRB1_0301", "HLA-DQA10501-DQB10201", "HLA-DQA10101-DQB10501", 
               "HLA-DPA10103-DPB10402", "HLA-DPA10103-DPB10401")
s_best <- df[is.element(df$variable, s_alleles), ]

phbrII <- c()
for (i in muts){
  values_or <- s_best$value[s_best$mut == i & s_best$state == "or"]
  values_mut <- s_best$value[s_best$mut == i & s_best$state == "mut"]
  if (sum(values_or <= 10) > 0 | sum(values_mut <= 10) > 0) {
    phbrII <- rbind(phbrII, c(i, 6/sum(1/values_or), 6/sum(1/values_mut), mutations$freq[mutations$name == i],
                          mutations$dyn[mutations$name == i], "yes"))
  } else {
    phbrII <- rbind(phbrII, c(i, 6/sum(1/values_or), 6/sum(1/values_mut), mutations$freq[mutations$name == i],
                          mutations$dyn[mutations$name == i], "no"))
  }
  #phbr <- rbind(phbr, c(i, 6/sum(1/values_mut), "mut"))
}
phbrII <- data.frame(phbrII)
colnames(phbrII) <- c("mut", "phbr_or", "phbr_mut", "freq", "dyn", "presented")
phbrII[,2:4] <- sapply(phbrII[,2:4], as.numeric)
phbrII$rel <- phbrII$phbr_mut/phbrII$phbr_or 

phbrII$phbr_mut[phbrII$mut == "ORF8:Q18*"] <- 1000
phbrII$rel[phbrII$mut == "ORF8:Q18*"] <- 14

phbrII$mut1 <- phbrII$mut
phbrII$mut1[phbrII$rel <= 2] <- NA

#phbrII$rel[phbrII$rel < 1] <- 1/phbrII$rel[phbrII$rel < 1]
p_mutII <- ggplot(phbrII[phbrII$presented == "yes" & phbrII$mut != "ORF8:Q18*",], aes(x = phbr_or, y = phbr_mut, label = mut1)) +
  geom_abline(slope = 1, linetype="dashed", color = "grey10") +
  geom_point(aes(fill = rel), size = 6, shape = 21, color = "grey20", alpha = 0.9) +
  scale_x_log10(limits = c(0.05, 30), name = "PHBR in ancestral state") +
  scale_y_log10(limits = c(0.05, 30), name = "PHBR in derived state") +
  scale_fill_distiller(name = "", palette = "RdYlBu", limits = c(0.19, 14)) +
  #ggtitle("HLA II") +
  geom_text_repel(size = 6) +
  theme_bw() +
   theme(legend.position = "none",
        axis.text.y = element_text(size = 18),
        axis.text.x = element_text(size = 18),
        legend.title = element_text (size = 18),
        legend.text = element_text(size = 18),
        axis.title = element_text (size = 18))
```

PHBRs of muts in general distribution
NOT INCLUDED

```{r}
distr1 <- readRDS(paste0("/home/jane/PhD_Skoltech/coronavirus/immunocompromised/", "phbr_distrI.Rds", collapse = ""))

plot_distr1 <- ggplot(distr1, aes(x = phbr)) +
  geom_density(alpha=0.4, fill = "lightsteelblue") +
  geom_vline(xintercept = phbrI$phbr_or[phbrI$presented == "yes" & phbrI$mut != "ORF8:Q18*"], linetype="dashed", color = "darkred") +
  scale_y_continuous("Density") +
  scale_x_continuous("PHBR on HLA I") +
  #scale_x_log10("PHBR on HLA I", limits = c(0.005, 200)) +
  #ggtitle("B Permutation test for HLA I") +
  #annotate("text", x = 7.5, y=0.04, label = paste0("p=0.0019"), size = 4) +
  theme_bw() +
                   theme(legend.position = "top", 
                   legend.background = element_rect(), 
                   legend.title = element_text (colour="black", size = 12, face = "plain"),
                   legend.text = element_text(colour="black", size = 12, face = "plain"),
                   plot.title = element_text(colour="black", size = 14, face = "bold"),
                   axis.title = element_text (colour="black", size = 14, face = "plain"),
                   axis.text.y = element_text(colour="grey20", size = 14, face = "plain"),
                   axis.text.x = element_text(colour="grey20", size = 14, face = "plain", angle = 90, vjust = 0.5),
                   axis.line = element_line(colour="grey20", size = 0.6),
                   panel.background = element_rect(fill = "white", colour = "grey20"),
                   strip.text = element_text(size = 12)) 
plot_distr1

sum(phbrI$phbr_or[phbrI$presented == "yes" & phbrI$mut != "ORF8:Q18*"] < median(distr1$phbr))
## 13
## 1.34 mean

p_distr1 <- c()
for (i in 1:10000){
  p_distr1 <- c(p_distr1, mean(sample(distr1$phbr, 27)))
}
p_value1 <- sum(mean(phbrI$phbr_or[phbrI$presented == "yes" & phbrI$mut != "ORF8:Q18*"]) < p_distr1)/length(p_distr1)
p_value1

distr2 <- readRDS(paste0("/home/jane/PhD_Skoltech/coronavirus/immunocompromised/", "phbr_distrII.Rds", collapse = ""))

plot_distr2 <- ggplot(distr2, aes(x = phbr)) +
  geom_density(alpha=0.4, fill = "lightsteelblue") +
  geom_vline(xintercept = phbrII$phbr_or[phbrII$presented == "yes" & phbrII$mut != "ORF8:Q18*"], linetype="dashed", color = "darkred") +
  scale_y_continuous("Density") +
  scale_x_continuous("PHBR on HLA II") +
  #scale_x_log10("PHBR on HLA II", limits = c(0.005, 200)) +
  #ggtitle("B Permutation test for HLA I") +
  #annotate("text", x = 7.5, y=0.04, label = paste0("p=0.0019"), size = 4) +
  theme_bw() +
                   theme(legend.position = "top", 
                   legend.background = element_rect(), 
                   legend.title = element_text (colour="black", size = 12, face = "plain"),
                   legend.text = element_text(colour="black", size = 12, face = "plain"),
                   plot.title = element_text(colour="black", size = 14, face = "bold"),
                   axis.title = element_text (colour="black", size = 14, face = "plain"),
                   axis.text.y = element_text(colour="grey20", size = 14, face = "plain"),
                   axis.text.x = element_text(colour="grey20", size = 14, face = "plain", angle = 90, vjust = 0.5),
                   axis.line = element_line(colour="grey20", size = 0.6),
                   panel.background = element_rect(fill = "white", colour = "grey20"),
                   strip.text = element_text(size = 12)) 
plot_distr2

sum(phbrII$phbr_or[phbrII$presented == "yes" & phbrII$mut != "ORF8:Q18*"] < median(distr2$phbr))
## 16
## mean 5.25
Mode <- function(x) {
  ux <- unique(x)
  ux[which.max(tabulate(match(x, ux)))]
}



p_distr2 <- c()
for (i in 1:10000){
  p_distr2 <- c(p_distr2, mean(sample(distr2$phbr, 24)))
}
p_value2 <- sum(mean(phbrII$phbr_or[phbrII$presented == "yes" & phbrII$mut != "ORF8:Q18*"]) < p_distr2)/length(p_distr2)
p_value2


plot <-  cowplot::plot_grid(plot_distr1, plot_distr2, rel_heights = c(1, 1), ncol = 1, labels = c("", ""))

pdf(file = "/home/jane/PhD_Skoltech/coronavirus/immunocompromised/plot_distr.pdf",  
    width = 6, # The width of the plot in inches
    height = 8)
plot
dev.off()
```


#############################################
MHC I affected epitopes are more affine
NOT INCLUDED

```{r}
df_or_MHC <- rbind(cbind(rep("HLA I", nrow(phbrI)), phbrI$mut, phbrI$dyn, phbrI$presented, phbrI$phbr_or, phbrI$rel),
                    cbind(rep("HLA II", nrow(phbrII)), phbrII$mut, phbrII$dyn, phbrII$presented, phbrII$phbr_or, phbrII$rel))

df_or_MHC <- data.frame(df_or_MHC)
colnames(df_or_MHC) <- c("hla", "mut", "dyn", "presented", "phbr", "rel")
df_or_MHC$phbr <- as.numeric(df_or_MHC$phbr)
df_or_MHC$mut1 <- df_or_MHC$mut
df_or_MHC$mut1[df_or_MHC$hla == "HLA I"] <- ""

p_box_or_all <- ggplot(df_or_MHC[df_or_MHC$mut != "ORF8:Q18*",], aes(x = hla, y = phbr, fill = as.numeric(rel), label = mut1)) +
  geom_boxplot(draw_quantiles = c(0.5), width = 0.35, fill = "grey70", color = "grey20", alpha = 0.8) +
  geom_line(aes(group = mut), size = 0.5) +
  geom_point(size = 5, shape = 21, color = "grey20", alpha = 0.8, position = position_dodge()) +
  #geom_text_repel( nudge_x = ifelse(exact$mut == "endornase:P205L", 0.7, 0.5),
  #                 max.iter = 1000, force = 5, segment.color = "grey60") +
  scale_fill_distiller(name = "", palette = "RdYlBu", limits = c(0.19, 14)) +
  scale_y_continuous(name = "PHBR", trans = "log10", limits = c(0.005, 120)) +
  scale_x_discrete(name = "", labels = c("HLA I" = "HLA I\n ", "HLA II" = "HLA II\n ")) +
  ggtitle("All changed sites") +
  stat_compare_means(paired = TRUE, size = 3.5) +
    theme_bw() +
  theme(legend.position = "none",
        axis.text.y = element_text(size = 14),
        axis.text.x = element_text(size = 14),
        legend.title = element_text (size = 14),
        legend.text = element_text(size = 14),
        axis.title = element_text (size = 14))
p_box_or_all
```

#####################################
Found specific for alleles

```{r}
found <- read.table("/home/jane/PhD_Skoltech/coronavirus/immunocompromised/found_june1.txt")
found <- data.frame(found)
colnames(found) <- c("allele", "peptide", "class")

gg_table1 <- readRDS("/home/jane/PhD_Skoltech/coronavirus/immunocompromised/net_pan/ggtable1_mhcI_Rank_EL.Rds")
gg_table1$cl <- rep("HLA I", nrow(gg_table1))

gg_table1$immuno <- rep("no", nrow(gg_table1))

found_i <- found$peptide[found$class == "HLAI"]
for (i in 1:length(found_i)){
  gg_table1$immuno[str_detect(gg_table1$pept_or, found_i[i])] <- "substr"
  #gg_table1$immuno[gg_table1$pept_or == found[i]] <- "exact"
}

for (i in 1:length(found_i)){
  #gg_table1$immuno[str_detect(gg_table1$pept_or, found[i])] <- "substr"
  gg_table1$immuno[gg_table1$pept_or == found_i[i]] <- "exact"
}
#gg_table1$immuno[as.numeric(gg_table1$Rank_or) > 2 & as.numeric(gg_table1$Rank_mut) > 2] <- "no"
muts_i <- unique(gg_table1$mut[gg_table1$immuno == "exact"])

gg_table2 <- readRDS("/home/jane/PhD_Skoltech/coronavirus/immunocompromised/net_pan/ggtable2_mhcII_Rank_EL.Rds")
gg_table2$cl <- rep("HLA II", nrow(gg_table2))

gg_table2$immuno <- rep("no", nrow(gg_table2))

found_ii <- found$peptide[found$class == "HLAII"]
for (i in 1:length(found_ii)){
  gg_table2$immuno[str_detect(gg_table2$pept_or, found_ii[i])] <- "substr"
  #if (sum(str_detect(gg_table2$pept_or, found_ii[i]))>0) {print(found_ii[i])}
  #gg_table1$immuno[gg_table1$pept_or == found[i]] <- "exact"
}

for (i in 1:length(found_ii)){
  #gg_table1$immuno[str_detect(gg_table1$pept_or, found[i])] <- "substr"
  gg_table2$immuno[gg_table2$pept_or == found_ii[i]] <- "exact"
}
#gg_table2$immuno[as.numeric(gg_table2$Rank_or) > 10 & as.numeric(gg_table2$Rank_mut) > 10] <- "no"
muts_ii <- unique(gg_table2$mut[gg_table2$immuno != "no"])
```


```{r}
phbrI$immuno <- as.numeric(is.element(phbrI$mut, muts_i))
phbrI$rel <- as.character(phbrI$rel)
df <- melt(phbrI[phbrI$immuno == "1", c(1,2,3,5,8)])
df <- df[df$mut != "ORF8:Q18*",]
df$mut1 <- df$mut
df$mut[df$variable == "phbr_or"] <- ""

p_boxI_immuno <- ggplot(df, aes(x = variable, y = value, label = mut)) +
  geom_boxplot(draw_quantiles = c(0.5), width = 0.35, fill = "grey70", color = "grey20", alpha = 0.8) +
  geom_line(aes(group = mut1)) +
  geom_point(aes(fill = as.numeric(df$rel)), size = 6, shape = 21, color = "grey20", alpha = 0.8, position = position_dodge()) +
 #geom_text_repel( nudge_x = ifelse(exact$mut == "endornase:P205L", 0.7, 0.5),
#                   max.iter = 1000, force = 5, segment.color = "grey60") +
  
  scale_x_discrete(labels=c("phbr_or" = "Ancestral", "phbr_mut" = "Derived"), name = "") +
  scale_y_continuous(name = "PHBR", trans = "log10") +
  scale_fill_distiller(name = "", palette = "RdYlBu", limits = c(0.19, 14)) +
  #ggtitle("Sites, immunogenic on HLA I") +
  #stat_compare_means(paired = TRUE, size = 4.5) +
  annotate("text", x = 1.2, y = 5, label = "p = 0.083", size = 5) +
  theme_bw() +
  theme(legend.position = "none",
        axis.text.y = element_text(size = 18),
        axis.text.x = element_text(size = 18),
        legend.title = element_text (size = 18),
        legend.text = element_text(size = 18),
        axis.title = element_text (size = 18))
p_boxI_immuno
```


```{r}
foundI <- found[found$class == "HLAI" & is.element(found$peptide, gg_table1$pept_or), ]

exact <- c()

for (i in muts_i[muts_i != "ORF8:Q18*" & muts_i != "E:F26L"]) {
  gg_i <- gg_table1[gg_table1$immuno == "exact" & gg_table1$mut == i,] 
  best_or <- c()
  best_mut <- c()
  for (j in 1:nrow(gg_i)){
    p_or <- gg_i$pept_or[j]
    p_mut <- gg_i$pept_mut[j]
    best_or <- c(best_or,
                 net_panI$Rank_EL[net_panI$peptide == p_or & net_panI$allele == foundI$allele[foundI$peptide == p_or]])
    best_mut <- c(best_mut,
                 net_panI$Rank_EL[net_panI$peptide == p_mut & net_panI$allele == foundI$allele[foundI$peptide == p_or]])
  }
 
  exact <- rbind(exact, c(i, gg_i$pept_or[which(best_or == min(best_or))], min(best_or), "or", 
                          foundI$allele[foundI$peptide == gg_i$pept_or[which(best_or == min(best_or))]]),
                c(i, gg_i$pept_mut[which(best_mut == min(best_mut))], min(best_mut), "mut", 
                          foundI$allele[foundI$peptide == gg_i$pept_or[which(best_mut == min(best_mut))]]))
}
exact <-data.frame(exact)
colnames(exact) <- c("mut", "pept", "rank", "state", "allele")
exact$rank <- as.numeric(exact$rank)

exact$mut1 <- exact$mut
exact$mut1[exact$state == "or"] <- ""

exact$state <- factor(exact$state, levels = c("or", "mut"))

exact$mut[exact$mut == "S:del141-144"] <- "S:del140_144"

rel <- c()
for (i in 1:nrow(exact)){
  rel <- c(rel, df_or_MHC$rel[df_or_MHC$mut == exact$mut[i] & df_or_MHC$hla == "HLA I"])
}
exact$rel <- rel

p_exact_rank <- ggplot(exact, aes(x = state, y = rank, fill = as.numeric(rel))) +
  annotate("rect", ymin=0, ymax=0.5, xmin = -Inf, xmax = Inf, 
           alpha=0.3, fill = "grey50", color="white") +
  annotate("rect", ymin=0.5, ymax=2, xmin = -Inf, xmax = Inf, 
           alpha=0.1, fill = "grey50", color="white") +
 
  #annotate("text", x = 0.65, y=0.1, label = 'bold("Strong")', parse = TRUE) +
  #annotate("text", x = 0.65, y=1, label = 'bold("Weak")', parse = TRUE) +
  #annotate("text", x = 0.65, y =10, label = 'bold("    No \nbinding")', parse = TRUE) +
  
  geom_hline(yintercept = 2, linetype="dashed", color = "grey20") +
  geom_hline(yintercept = 0.5, linetype="dashed", color = "grey20") +
  geom_boxplot(draw_quantiles = c(0.5), width = 0.35,  color = "grey20", alpha = 0.8) +
  geom_line(aes(group = mut)) +
  geom_point(color = "grey20", size = 6, shape = 21, alpha = 1, position = position_dodge()) +
  
  #geom_text_repel( nudge_x = ifelse(exact$mut == "endornase:P205L", 0.7, 0.5), nudge_y = - 2,
  #                 max.iter = 1000, force = 5, segment.color = "grey60", size = 3.5) +
  #geom_text(x = 5.5, size = 5) +
  scale_fill_distiller(name = "", palette = "RdYlBu", limits = c(0.19, 14)) +
  scale_y_continuous(name = "imBR", trans = "log10", limits = c(0.001, 100)) +
  scale_x_discrete(labels=c("or" = "Ancestral", "mut" = "Derived"), name = "") +
  #stat_compare_means(paired = TRUE, size = 4.5) +
  annotate("text", x = 1.2, y =70, label = "p = 0.00098", size = 5) +
  guides(color = FALSE, size = FALSE, alpha = FALSE) +
  #ggtitle("Sites, immunogenic on HLA I") +
                                                  # Annotate text
    theme_bw() +
                   theme(legend.position = "none", 
                   legend.background = element_rect(), 
                   legend.title = element_text (colour="black", size = 18, face = "plain"),
                   legend.text = element_text(colour="black", size = 18, face = "plain"),
                   plot.title = element_text(colour="black", size = 18, face = "plain"),
                   axis.title = element_text (colour="black", size = 18, face = "plain"),
                   axis.text.y = element_text(colour="grey20", size = 18, face = "plain"),
                   axis.text.x = element_text(colour="grey20", size = 18, face = "plain"),
                   axis.line = element_line(colour="grey20", size = 0.6),
                   panel.background = element_rect(fill = "white", colour = "grey20"),
                   strip.text = element_text(size = 18)) 
  #theme(plot.margin = unit(c(1, 5, 1, 1), "lines")) +
  #coord_cartesian(clip = "off")
p_exact_rank
```

First pannel

```{r}

# Draw text
#::::::::::::::::::::::::::::::::::::::
text <- paste("", "", "", "S:del141-144", "N:P6T", "M:L129R", "endornase:P205L", "", "S:S50L", "nsp3:T1456I",
              "nsp3:T504P", "S:R273S", "S:T470", "nsp3:T504A", "nsp3:D821N", "", "", "", "", "", "", "", sep = "\n")


text.p <- ggparagraph(text, face = "bold", size = 12)
text.p

setwd("/home/jane/PhD_Skoltech/coronavirus/immunocompromised/")
PostScriptTrace("sch1.eps")
sch1 <- readPicture("sch1.eps.xml")
sch1 <- pictureGrob(sch1)


p1 <- cowplot::plot_grid(sch1)


legend <- cowplot::get_legend(p_mutI + theme(legend.position = "left"))
p2 <- cowplot::plot_grid(p_mutI, p_mutII, rel_widths = c(1, 1), ncol = 2,  labels = c("B", ""))
p3 <- cowplot::plot_grid(legend, p_boxI_immuno, p_exact_rank, text.p, rel_widths = c(0.2, 1, 1, 0.4), ncol = 4,  
                          labels = c("", "C", "D", ""))

plot1 <-  cowplot::plot_grid(NULL, p1, NULL, p2, p3, rel_heights = c(-0.1, 1.1, -0.08, 1.1, 0.9), ncol = 1, labels = c("A", ""))

pdf(file = "/home/jane/PhD_Skoltech/coronavirus/immunocompromised/plot1_phbr.pdf",  
    width = 12, # The width of the plot in inches
    height = 14)
plot1
dev.off()
```

UGLY pannel

```{r}
text <- paste("", "", "", "S:del141-144", "N:P6T", "M:L129R", "endornase:P205L", "", "S:S50L", "nsp3:T1456I",
              "nsp3:T504P", "S:R273S", "S:T470", "nsp3:T504A", "nsp3:D821N", "", "", "", "", "", "", "", sep = "\n")


text.p <- ggparagraph(text, face = "bold", size = 12)
text.p

setwd("/home/jane/PhD_Skoltech/coronavirus/immunocompromised/")
PostScriptTrace("sch1_edit.eps")
sch1 <- readPicture("sch1_edit.eps.xml")
sch1 <- pictureGrob(sch1)


p1 <- cowplot::plot_grid(NaN, sch1, NaN, rel_heights = c(-0.18, 1, -0.05), ncol = 1)
#p_box <- cowplot::plot_grid(p_boxI_immuno, NaN, rel_heights = c(1, 0.2), ncol = 1)
#p1 <- cowplot::plot_grid(p1, p_box, rel_widths = c(1.2, 0.4), ncol = 2, labels = c("A", "D"), label_size = 20)


legend <- cowplot::get_legend(p_mutI + theme(legend.position = "left"))
p2 <- cowplot::plot_grid(legend, p_mutI, NaN, p_mutII,  p_exact_rank, rel_widths = c(0.14, 0.55, 0.01, 0.55, 0.36), ncol = 5,  
                         labels = c("", "b", "", "c", "d"), label_size = 20)

plot1 <-  cowplot::plot_grid( p1, NaN, p2, rel_heights = c(1.1, -0.2, 1), labels = c("a", ""), label_size = 20, ncol = 1)

pdf(file = "/home/jane/PhD_Skoltech/coronavirus/immunocompromised/plot1_phbr_ugly.pdf",  
    width = 15, # The width of the plot in inches
    height = 8)
plot1
dev.off()

```



6.83 - 6.01


```{r}
foundII <- found[found$class == "HLAII", ]

exact2 <- c()
##  
for (i in muts_ii[muts_ii != "ORF8:Q18*"]) {
  gg_i <- gg_table2[gg_table2$immuno != "no" & gg_table2$mut == i,] 
  best_or <- c()
  best_mut <- c()
  for (j in 1:nrow(gg_i)){
    p_or <- gg_i$pept_or[j]
    p_mut <- gg_i$pept_mut[j]
    best_or <- c(best_or,
                 net_pan$Rank_EL[net_pan$peptide == p_or & net_pan$allele == foundII$allele[foundII$peptide == p_or]])
    best_mut <- c(best_mut,
                 net_pan$Rank_EL[net_pan$peptide == p_mut & net_pan$allele == foundII$allele[foundII$peptide == p_or]])
  }
 
  exact2 <- rbind(exact2, c(i, gg_i$pept_or[which(best_or == min(best_or))], min(best_or), "or", 
                          foundII$allele[foundII$peptide == gg_i$pept_or[which(best_or == min(best_or))]]),
                c(i, gg_i$pept_mut[which(best_mut == min(best_mut))], min(best_mut), "mut", 
                          foundII$allele[foundII$peptide == gg_i$pept_mut[which(best_mut == min(best_mut))]]))
}
exact2 <-data.frame(exact2)
colnames(exact2) <- c("mut", "pept", "rank", "state", "allele")
exact2$rank <- as.numeric(exact2$rank)



```

```{r}
setwd("/home/jane/PhD_Skoltech/coronavirus/immunocompromised/")
PostScriptTrace("fig2_final_tree_main.eps")
tree <- readPicture("fig2_final_tree_main.eps.xml")
tree <- pictureGrob(tree)

```



########################################3
Population
надо делить

```{r}
path = "/home/jane/PhD_Skoltech/coronavirus/immunocompromised/net_pan/"
best_rank_alleles <- readRDS(paste0(path, "best_rank_mhcI.Rds", collapse = ""))
best_rank_alleles %>% 
  melt -> df
s_alleles <- c("HLA-A*01:01", "HLA-A*03:01", "HLA-B*07:02", "HLA-B*08:01", "HLA-C*07:01", "HLA-C*07:02")

dfI <- cbind(df[df$state == "or",], df$value[df$state == "mut"], df$value[df$state == "dev"])
colnames(dfI) <- c("mut", "state", "allele", "rank_or", "rank_mut", "dev")
dfI <- dfI[dfI$rank_or <= 2 | dfI$rank_mut <= 2, ]

dfI$family <- str_extract(dfI$allele, "HLA-[A-C]*")
dfI$s <- as.numeric(is.element(dfI$allele, s_alleles))
dfI$s <- factor(dfI$s, levels = c("1", "0"))
dfI <- dfI[order(1/(as.numeric(dfI$s) + 1)), ]
dfI <- dfI[dfI$mut != "ORF7b:del2",]

immunogenic <- c("S:S50L", "S:del140_144", "S:R273S", "S:T470N", "nsp3:T504A", "nsp3:T504P", "nsp3:D821N", "nsp3:T1456I",
                 "endornase:P205L", "N:P6T", "M:L129R")

av <- c()
n=100000
for (i in 1:n){
  hla_sample <- c()
  for (f in unique(dfI$family)){
    hla_sample <- c(hla_sample, as.character(sample(unique(dfI$allele[dfI$family == f]), 2)))
  }
  sample_av <- mean(dfI$dev[is.element(dfI$mut, immunogenic) & is.element(dfI$allele, hla_sample)])
  av <- c(av, sample_av)
}
real_av <- mean((dfI$dev[dfI$s == 1 & is.element(dfI$mut, immunogenic)]))
p_valueI <- 1 - sum(real_av >= av)/n 

av <- data.frame(av)

dfI$mut <- factor(dfI$mut, levels = c("nsp2:A411V", "nsp3:T504A", "nsp3:T504P", "nsp3:D821N",
  "nsp3:T1456I", "nsp4:T295I", "nsp4:V315I", "nsp6:del37_37", "nsp6:L37F", "nsp6:V278I", "nsp7:V58G", 
  "RdRp:N158S", "endornase:P205L", "S:S50L", "S:del68_70", "S:del140_144", "S:R273S", "S:N440K", "S:L452R",
  "S:Y453F", "S:T470N", "S:G476S", "S:P621S", "S:D737G", "E:S6L", "E:F26L", "M:L129R", "ORF7a:del2_2",
  "ORF8:Q18*",  "N:P6T", "N:R195G"))

dfI$hla_l <- substring(dfI$allele, 5, 11)
dfI$hla_l[dfI$s != "1" | dfI$dev <= 5] = NA
p_mI <- ggplot(dfI[dfI$mut != "ORF8:Q18*",], aes(x = mut, y = dev, fill = as.factor(s), 
                                                 size = as.factor(s), label = hla_l)) +
  annotate("rect", xmin=1.5, xmax=5.5, ymin = 0, ymax = Inf, 
           alpha=0.4, fill = "wheat", color="white") +
  annotate("rect", xmin=12.5, xmax=14.5, ymin = 0, ymax = Inf, 
           alpha=0.4, fill = "wheat", color="white") +
   annotate("rect", xmin=15.5, xmax=17.5, ymin = 0, ymax = Inf, 
           alpha=0.4, fill = "wheat", color="white") +
  annotate("rect", xmin=20.5, xmax=21.5, ymin = 0, ymax = Inf, 
           alpha=0.4, fill = "wheat", color="white") +
  annotate("rect", xmin=26.5, xmax=27.5, ymin = 0, ymax = Inf, 
           alpha=0.4, fill = "wheat", color="white") +
  annotate("rect", xmin=28.5, xmax=29.5, ymin = 0, ymax = Inf, 
           alpha=0.4, fill = "wheat", color="white") +
  
  geom_hline(yintercept = 1, linetype="dashed", color = "grey20") +
  scale_y_continuous(trans = "log10", name = "BR fold change") +
  scale_x_discrete(name = "") +
  geom_quasirandom(shape = 21, color = "black", alpha = 0.7) +
  geom_text_repel(aes(label = hla_l), vjust = 0.5, size = 6) +
  #geom_text(aes(label= p_extr, y = 100),  size=5,  position = position_dodge(0.9), color = "grey10", face = "bold")+
  scale_fill_manual(name = "Allele", values = c("0" = "lightsteelblue3", "1" = "darkred"), 
                    labels = c("0" = "Other", "1" = "Patient S")) +
  scale_size_manual(values = c(4, 2.5)) +
  #ggtitle("A  HLA I") +
  guides(size = FALSE,
         fill = guide_legend(override.aes = list(size = 4))) +
 theme_bw() +
                   theme(legend.position = "none", 
                   legend.background = element_rect(), 
                   legend.title = element_text (colour="black", size = 18, face = "plain"),
                   legend.text = element_text(colour="black", size = 18, face = "plain"),
                   plot.title = element_text(colour="black", size = 18, face = "bold"),
                   axis.title = element_text (colour="black", size = 18, face = "plain"),
                   axis.text.y = element_text(colour="grey20", size = 18, face = "plain"),
                   axis.text.x = element_text(colour="grey20", size = 18, face = "plain", angle = 30, vjust = 1, hjust = 1),
                   axis.line = element_line(colour="grey20", size = 0.6),
                   panel.background = element_rect(fill = "white", colour = "grey20"),
                   strip.text = element_text(size = 18))  
p_mI
```


```{r}
p_statI <- ggplot(av, aes(x = av)) + 
  geom_density(alpha=0.4, fill = "wheat") +
  geom_vline(xintercept = real_av, linetype="dashed", color = "darkred") +
  scale_y_continuous("Density") +
  scale_x_continuous("Mean BR fold change") +
  #ggtitle("B Permutation test for HLA I") +
  annotate("text", x = 7, y=0.1, label = "p=0.0033", size = 6) +
  theme_bw() +
                  theme(legend.position = "none", 
                   legend.background = element_rect(), 
                   legend.title = element_text (colour="black", size = 18, face = "plain"),
                   legend.text = element_text(colour="black", size = 18, face = "plain"),
                   plot.title = element_text(colour="black", size = 18, face = "bold"),
                   axis.title = element_text (colour="black", size = 18, face = "plain"),
                   axis.text.y = element_text(colour="grey20", size = 18, face = "plain"),
                   axis.text.x = element_text(colour="grey20", size = 18, face = "plain", angle = 30, vjust = 0.5),
                   axis.line = element_line(colour="grey20", size = 0.6),
                   panel.background = element_rect(fill = "white", colour = "grey20"),
                   strip.text = element_text(size = 18)) 
p_statI
```


```{r}
path = "/home/jane/PhD_Skoltech/coronavirus/immunocompromised/net_pan/"
best_rank_alleles <- readRDS(paste0(path, "best_rank_mhcII.Rds", collapse = ""))
best_rank_alleles %>% 
  melt -> df

s_allelesII <- c("DRB1_0101",  "DRB1_0301", "HLA-DQA10501-DQB10201", "HLA-DQA10101-DQB10501", 
               "HLA-DPA10103-DPB10402", "HLA-DPA10103-DPB10401")

dfII <- cbind(df[df$state == "or",], df$value[df$state == "mut"], df$value[df$state == "dev"])
colnames(dfII) <- c("mut", "state", "allele", "rank_or", "rank_mut", "dev")
dfII <- dfII[dfII$rank_or <= 10 | dfII$rank_mut <= 10, ]

dfII$family <- str_extract(dfII$allele, "D[A-Z]*")
dfII$s <- as.numeric(is.element(dfII$allele, s_allelesII))
dfII$s <- factor(dfII$s, levels = c("1", "0"))
dfII <- dfII[order(1/(as.numeric(dfII$s) + 1)), ]
dfII <- dfII[dfII$mut != "ORF7b:del2",]
dfII$mut <- factor(dfII$mut, levels = c("nsp2:A411V", "nsp3:T504A", "nsp3:T504P", "nsp3:D821N",
  "nsp3:T1456I", "nsp4:T295I", "nsp4:V315I", "nsp6:del37_37", "nsp6:L37F", "nsp6:V278I", "nsp7:V58G", 
  "RdRp:N158S", "endornase:P205L", "S:S50L", "S:del68_70", "S:del140_144", "S:R273S", "S:N440K", "S:L452R",
  "S:Y453F", "S:T470N", "S:G476S", "S:P621S", "S:D737G", "E:S6L", "E:F26L", "M:L129R", "ORF7a:del2_2",
  "ORF8:Q18*",  "N:P6T", "N:R195G"))



p_mII <- ggplot(dfII[dfII$mut != "ORF8:Q18*",], aes(x = mut, y = dev, fill = as.factor(s), size = as.factor(s))) +
  annotate("rect", xmin=13.5, xmax=14.5, ymin = 0, ymax = Inf, 
           alpha=0.4, fill = "wheat", color="white") +
  geom_hline(yintercept = 1, linetype="dashed", color = "grey20") +
  scale_y_continuous(trans = "log10", name = "BR fold change") +
  scale_x_discrete(name = "", position = "bottom") +
  geom_quasirandom(shape = 21, color = "black", alpha = 0.7) +
  #geom_text(aes(label= p_extr, y = 100),  size=5,  position = position_dodge(0.9), color = "grey10", face = "bold")+
  scale_fill_manual(name = "Allele:", values = c("0" = "lightsteelblue3", "1" = "darkred"), 
                    labels = c("0" = "Other", "1" = "Patient S")) +
  scale_size_manual(values = c(4, 2.5)) +
  #ggtitle("C HLA II") +
  guides(size = FALSE,
         fill = guide_legend(override.aes = list(size = 4))) +
  theme_bw() +
                   theme(legend.position = "none", 
                   legend.background = element_rect(), 
                   legend.title = element_text (colour="black", size = 18, face = "plain"),
                   legend.text = element_text(colour="black", size = 18, face = "plain"),
                   plot.title = element_text(colour="black", size = 18, face = "bold"),
                   axis.title = element_text (colour="black", size = 18, face = "plain"),
                   axis.text.y = element_text(colour="grey20", size = 18, face = "plain"),
                   axis.text.x = element_text(colour="grey20", size = 18, face = "plain", angle = 30, vjust = 1, hjust = 1),
                   axis.line = element_line(colour="grey20", size = 0.6),
                   panel.background = element_rect(fill = "white", colour = "grey20"),
                   strip.text = element_text(size = 18)) 
p_mII
```


```{r}
immunogenicII <- c("S:S50L")

avII <- c()
n=100000
for (i in 1:n){
  hla_sample <- c()
  for (f in unique(dfII$family)){
    hla_sample <- c(hla_sample, as.character(sample(unique(dfII$allele[dfII$family == f]), 2)))
  }
  sample_av <- mean(dfII$dev[is.element(dfII$mut, immunogenicII) & is.element(dfII$allele, hla_sample)])
  avII <- c(avII, sample_av)
}
real_avII <- mean((dfII$dev[dfII$s == 1 & is.element(dfII$mut, immunogenicII)]))
p_valueII <- 1 - sum(real_avII >= avII)/n 

avII <- data.frame(avII)

p_statII <- ggplot(avII, aes(x = avII)) + 
  geom_density(alpha=0.4, fill = "wheat") +
  geom_vline(xintercept = real_avII, linetype="dashed", color = "darkred") +
  scale_y_continuous("Density") +
  scale_x_continuous("Mean BR fold change", limits = c(0, 8)) +
  #ggtitle("D Permutation test for HLA II") +
  annotate("text", x = 3.5, y=0.5, label = paste0("p=0.6996"), size = 6) +
  theme_bw() +
                   theme(legend.position = "none", 
                   legend.background = element_rect(), 
                   legend.title = element_text (colour="black", size = 18, face = "plain"),
                   legend.text = element_text(colour="black", size = 18, face = "plain"),
                   plot.title = element_text(colour="black", size = 18, face = "bold"),
                   axis.title = element_text (colour="black", size = 18, face = "plain"),
                   axis.text.y = element_text(colour="grey20", size = 18, face = "plain"),
                   axis.text.x = element_text(colour="grey20", size = 18, face = "plain", angle = 30, vjust = 0.5),
                   axis.line = element_line(colour="grey20", size = 0.6),
                   panel.background = element_rect(fill = "white", colour = "grey20"),
                   strip.text = element_text(size = 18)) 
p_statII
```

```{r}
df_leg2 <- cbind(1:2, 1:2, c("immunogenic", "not found in IEDB"))
df_leg2 <- data.frame(df_leg2)
colnames(df_leg2) <- c("x1", "y1", "cell")

leg2_plot <- ggplot(df_leg2, aes(x = as.numeric(x1), y = as.numeric(y1), fill = cell)) +
  geom_point(size = 7, color = "black", shape = 22, alpha = 0.4) +
  scale_fill_manual("Position", values = c("immunogenic" = "wheat", "not found in IEDB" = "white")) + 
  theme_bw() +
  theme(legend.position = "left")



legend <- cowplot::get_legend(p_mI + theme(legend.position = "left",
                              legend.title = element_text (colour="black", size = 14, face = "bold"),
                              legend.text = element_text(colour="black", size = 14, face = "plain")))
legend2 <- cowplot::get_legend(leg2_plot + theme(legend.position = "left",
                              legend.title = element_text (colour="black", size = 14, face = "bold"),
                              legend.text = element_text(colour="black", size = 14, face = "plain")))
legend <- cowplot::plot_grid(NaN, legend, legend2, NaN, rel_widths = c(0.1, 1, 2, 0.9), ncol = 4)
p_m <- cowplot::plot_grid(legend, p_mII, rel_heights = c(0.5, 5), ncol = 1)
p1 <- cowplot::plot_grid(p_mI, NaN, p_m, rel_widths = c(1, 0.09, 0.98), ncol = 3)
p2 <- cowplot::plot_grid(NaN, p_statI, p_statII, NaN, rel_widths = c(2.2, 8.2, 8.1, 3.1), ncol = 4)


plot2 <-  cowplot::plot_grid(p1, p2, rel_heights = c(2.8, 1), ncol = 1)

pdf(file = "/home/jane/PhD_Skoltech/coronavirus/immunocompromised/plot2_phbr.pdf",  
    width = 12, # The width of the plot in inches
    height = 12)
plot2
dev.off()
```

```{r}
dfI$p_av <- rep("", nrow(dfI))
dfI$p_extr <- rep("", nrow(dfI))

sign <- c("ns", ".", "*", "**", "***")
sign_thr <- c(1, 0.1, 0.05, 0.01, 0.001)
for (i in muts){
  hla_values <- df$value[df$mut == i]
  hla_real <- df$value[df$mut == i & is.element(df$variable, s_alleles)]
  av_distr <- c()
  extr_distr <- c()
  for (j in 1:1000){
    hla_sample <- sample(hla_values, 6, replace = FALSE)
    av_distr <- c(av_distr, mean(hla_sample))
    extr_distr <- c(extr_distr, max(hla_sample))
    
  }
  p_av <- sum(av_distr >= mean(hla_real)) / 1000
  p_extr <- sum(extr_distr >= max(hla_real)) / 1000
  
  df$p_av[df$mut == i & df$variable == "HLA-A*01:01"] <- sign[sum(sign_thr >= p_av)]
  df$p_extr[df$mut == i & df$variable == "HLA-A*01:01"] <- sign[sum(sign_thr >= p_extr)]
}


```

#####
Mean rank before mean rank after

```{r}
path <- "/home/jane/PhD_Skoltech/coronavirus/immunocompromised/net_pan/"
al_effect1 <- readRDS(paste0(path, "al_effect_hlaI_Rank_EL_nostop.Rds", collapse = ""))
alleles <- unique(al_effect1$allele)
df_pop <- c()
for (al in alleles){
  or <- mean(dfI$rank_or[dfI$allele == al & dfI$mut != "ORF8:Q18*"])
  mut <- mean(dfI$rank_mut[dfI$allele == al & dfI$mut != "ORF8:Q18*"])
  freq <- al_effect1$freq[al_effect1$allele == al]
  df_pop <- rbind(df_pop, 
  c(al, or, mut, freq))
}

df_pop <- data.frame(df_pop)
colnames(df_pop) <- c("al", "or", "mut", "freq")
df_pop$s <- as.numeric(is.element(df_pop$al, s_alleles))
df_pop$s <- factor(df_pop$s, levels = c("1", "0"))
df_pop <- df_pop[order(1/(as.numeric(df_pop$s) + 1)), ]
df_pop[, 2:4] <- sapply(df_pop[,2:4], as.numeric)
df_pop$fam <- substring(df_pop$al, 1, 5)
df_pop$lab <- substring(df_pop$al, 7, 11)

p_dots_I <- ggplot(df_pop, aes(x = or, y = mut, fill = as.factor(s), label = lab)) +
  geom_abline(slope = 1, linetype="dashed", color = "grey10") +
  geom_point(aes(size = freq), shape = 21, color = "black", alpha = 0.7) +
  scale_x_log10(name = "Mean BR of ancestral", limits = c(0.3, 5)) +
  scale_y_log10(name = "Mean BR of derived", limits = c(0.3, 5)) +
  geom_text_repel(size=6, color = "grey10", face = "bold")+
  scale_fill_manual(name = "Allele", values = c("0" = "lightsteelblue4", "1" = "darkred"), 
                    labels = c("0" = "Other", "1" = "Patient S")) +
  scale_size_continuous(name = "Population\nfrequency", range = c(2, 8)) +
  guides(size = guide_legend(override.aes = list(fill = "lightsteelblue4")),
         fill = guide_legend(override.aes = list(size = 4))) +
  theme_bw() +
                   theme(legend.position = "none", 
                   legend.background = element_rect(), 
                   legend.title = element_text (colour="black", size = 18, face = "plain"),
                   legend.text = element_text(colour="black", size = 18, face = "plain"),
                   plot.title = element_text(colour="black", size = 18, face = "bold"),
                   axis.title = element_text (colour="black", size = 18, face = "plain"),
                   axis.text.y = element_text(colour="grey20", size = 18, face = "plain"),
                   axis.text.x = element_text(colour="grey20", size = 18, face = "plain", angle = 0, vjust = 0.5),
                   axis.line = element_line(colour="grey20", size = 0.6),
                   panel.background = element_rect(fill = "white", colour = "grey20"),
                   strip.text = element_text(size = 18))  +
  facet_wrap(~fam, ncol = 3)

p_dots_I

```

```{r}
path <- "/home/jane/PhD_Skoltech/coronavirus/immunocompromised/net_pan/"
al_effect2 <- readRDS(paste0(path, "al_effect_hlaII_Rank_EL_nostop.Rds", collapse = ""))
allelesII <- unique(al_effect2$allele)

df_pop2 <- c()
for (al in allelesII){
  or <- mean(dfII$rank_or[dfII$allele == al & dfII$mut != "ORF8:Q18*"])
  mut <- mean(dfII$rank_mut[dfII$allele == al & dfII$mut != "ORF8:Q18*"])
  freq <- al_effect2$freq[al_effect2$allele == al]
  df_pop2 <- rbind(df_pop2, 
  c(al, or, mut, freq, al_effect2$al_group[al_effect2 == al], substring(al_effect2$nice[al_effect2 == al], 6, 10)))
}

df_pop2 <- data.frame(df_pop2)
colnames(df_pop2) <- c("al", "or", "mut", "freq", "fam", "lab")
df_pop2$s <- as.numeric(is.element(df_pop2$al, s_allelesII))
df_pop2$s <- factor(df_pop2$s, levels = c("1", "0"))
df_pop2 <- df_pop2[order(1/(as.numeric(df_pop2$s) + 1)), ]
df_pop2[, 2:4] <- sapply(df_pop2[,2:4], as.numeric)

p_dots_II <- ggplot(df_pop2, aes(x = or, y = mut, fill = as.factor(s), label = lab)) +
  geom_abline(slope = 1, linetype="dashed", color = "grey10") +
  geom_point(aes(size = freq), shape = 21, color = "black", alpha = 0.7) +
  scale_x_log10(name = "Mean BR of ancestral", limits = c(3, 8)) +
  scale_y_log10(name = "Mean BR of derived", limits = c(3, 8)) +
  geom_text_repel(size=6, color = "grey10", face = "bold")+
  scale_fill_manual(name = "Allele", values = c("0" = "lightsteelblue4", "1" = "darkred"), 
                    labels = c("0" = "Other", "1" = "Patient S")) +
  scale_size_continuous(name = "Population\nfrequency", range = c(2, 8)) +
  guides(size = guide_legend(override.aes = list(fill = "lightsteelblue4")),
         fill = FALSE) +
  theme_bw() +
                    theme(legend.position = "none", 
                   legend.background = element_rect(), 
                   legend.title = element_text (colour="black", size = 18, face = "plain"),
                   legend.text = element_text(colour="black", size = 18, face = "plain"),
                   plot.title = element_text(colour="black", size = 18, face = "bold"),
                   axis.title = element_text (colour="black", size = 18, face = "plain"),
                   axis.text.y = element_text(colour="grey20", size = 18, face = "plain"),
                   axis.text.x = element_text(colour="grey20", size = 18, face = "plain", angle = 0, vjust = 0.5),
                   axis.line = element_line(colour="grey20", size = 0.6),
                   panel.background = element_rect(fill = "white", colour = "grey20"),
                   strip.text = element_text(size = 18))  +
  facet_wrap(~fam, ncol = 3)

p_dots_II

```

Join third plot

```{r}
p3 <- cowplot::plot_grid(p_dots_I, p_dots_II, rel_heights = c(1, 1), ncol = 1,  labels = c("A", "B"))

pdf(file = "/home/jane/PhD_Skoltech/coronavirus/immunocompromised/plot3_br.pdf",  
    width = 12, # The width of the plot in inches
    height = 9)
p3
dev.off()


```

#####
Ugly pannel

```{r}
df_leg2 <- cbind(1:2, 1:2, c("immunogenic", "not found in IEDB"))
df_leg2 <- data.frame(df_leg2)
colnames(df_leg2) <- c("x1", "y1", "cell")

leg2_plot <- ggplot(df_leg2, aes(x = as.numeric(x1), y = as.numeric(y1), fill = cell)) +
  geom_point(size = 7, color = "black", shape = 22, alpha = 0.4) +
  scale_fill_manual("Position", values = c("immunogenic" = "wheat", "not found in IEDB" = "white")) + 
  theme_bw() +
  theme(legend.position = "left",  legend.background = element_rect(), 
                   legend.title = element_text (colour="black", size = 18, face = "plain"),
                   legend.text = element_text(colour="black", size = 18, face = "plain"))



legend <- cowplot::get_legend(p_mI + theme(legend.position = "top",
                              legend.title = element_text (colour="black", size = 18, face = "bold"),
                              legend.text = element_text(colour="black", size = 18, face = "plain")))
legend2 <- cowplot::get_legend(leg2_plot + theme(legend.position = "top",
                              legend.title = element_text (colour="black", size = 18, face = "bold"),
                              legend.text = element_text(colour="black", size = 18, face = "plain")))
legend <- cowplot::plot_grid(NaN, legend, legend2, NaN, rel_widths = c(1, 1, 1.5, 0.5), nrow = 1)

p1 <- cowplot::plot_grid(p_mI, NaN, p_mII, NaN, rel_heights = c(1, -0.09, 1, -0.09), ncol = 1, labels = c("a","", "c"),
                         label_size = 20)
p2 <- cowplot::plot_grid(p_statI, NaN, p_statII, NaN, rel_heights = c( 0.9, 0.1, 0.88, 0.12), ncol = 1, 
                         labels = c("b", "", "d", ""), label_size = 20)
p3 <- cowplot::plot_grid(p_dots_I, p_dots_II, rel_widths = c(1, 1), nrow = 1, labels = c("e", "f"), label_size = 20)

p4 <-  cowplot::plot_grid(p1, p2,  rel_widths = c(4.2, 1), nrow = 1)
plot2 <-  cowplot::plot_grid(legend, p4, p3,  rel_heights = c(0.1, 2, 1), ncol = 1)

pdf(file = "/home/jane/PhD_Skoltech/coronavirus/immunocompromised/plot2_phbr_ugly.pdf",  
    width = 18, # The width of the plot in inches
    height = 12)
plot2
dev.off()
```




#### 
Heatmaps

```{r}
p_hm <- ggplot(df[str_detect(df$variable, "HLA-C"),], aes(x = mut, variable)) +
  geom_tile(aes(fill = value)) +
  scale_fill_distiller(name = "log(After/Before)", palette = "RdBu", trans = "log10") +
  theme_bw() +
                   theme(legend.position = "right", 
                   legend.background = element_rect(), 
                   legend.title = element_text (colour="black", size = 12, face = "plain"),
                   legend.text = element_text(colour="black", size = 12, face = "plain"),
                   plot.title = element_text(colour="black", size = 14, face = "bold"),
                   axis.title = element_text (colour="black", size = 14, face = "plain"),
                   axis.text.y = element_text(colour="grey20", size = 14, face = "plain"),
                   axis.text.x = element_text(colour="grey20", size = 14, face = "plain", angle = 90, vjust = 0.5),
                   axis.line = element_line(colour="grey20", size = 0.6),
                   panel.background = element_rect(fill = "white", colour = "grey20"),
                   strip.text = element_text(size = 12)) 
p_hm
  
```

```{r}
best_rank_alleles <- best_rank_alleles[best_rank_alleles$mut != "ORF8:Q18*",]
df_av <- c()
for (i in 3:97) {
       df_av <- rbind(df_av,
                   c(colnames(best_rank_alleles)[i], 
                     mean(best_rank_alleles[best_rank_alleles$state == "or",i]), 
                     mean(best_rank_alleles[best_rank_alleles$state == "mut",i]), 
                     1/mean(1/best_rank_alleles[best_rank_alleles$state == "mut",i]),
                     1/mean(1/best_rank_alleles[best_rank_alleles$state == "mut",i])))
    
  
}
df_av <- data.frame(df_av)
colnames(df_av) <- c("allele", "av_or", "harm_or",  "av_mut", "harm_mut")
s_alleles <- c("HLA-A*01:01", "HLA-A*03:01", "HLA-B*07:02", "HLA-B*08:01", "HLA-C*07:01", "HLA-C*07:02")
df_av$s <- as.numeric(is.element(df_av$allele, s_alleles))
df_av$family <- str_extract(df_av$allele, "HLA-[A-C]*")
df_av$lab <- substring(df_av$allele, 7, 11)

ggplot(df_av, aes(x = as.numeric(av_or), y = as.numeric(av_mut), fill = as.factor(s), label = lab)) +
  geom_abline(slope = 1, linetype="dashed", color = "grey10") +
  geom_point(shape = 21, color = "black", alpha = 0.7) +
  scale_x_continuous(limits = c(0, 9)) +
  scale_y_continuous(limits = c(0, 9)) +
  geom_text_repel(size=5, color = "grey10", face = "bold")+
  scale_fill_manual(name = "Allele:", values = c("0" = "lightsteelblue3", "1" = "darkred"), 
                    labels = c("0" = "Other", "1" = "Patient S")) +
 
  guides(#size = FALSE,
         fill = guide_legend(override.aes = list(size = 4))) +
  theme_bw() +
                   theme(legend.position = "none", 
                   legend.background = element_rect(), 
                   legend.title = element_text (colour="black", size = 12, face = "plain"),
                   legend.text = element_text(colour="black", size = 12, face = "plain"),
                   plot.title = element_text(colour="black", size = 14, face = "bold"),
                   axis.title = element_text (colour="black", size = 14, face = "plain"),
                   axis.text.y = element_text(colour="grey20", size = 14, face = "plain"),
                   axis.text.x = element_text(colour="grey20", size = 14, face = "plain", angle = 0, vjust = 0.5),
                   axis.line = element_line(colour="grey20", size = 0.6),
                   panel.background = element_rect(fill = "white", colour = "grey20"),
                   strip.text = element_text(size = 12)) +
  facet_wrap(~family)

```


######
Presentation score distributions for all peptides
mhcI - 8-11, mhcII - 12-18

```{r}
neutr <- read.table("/home/jane/PhD_Skoltech/coronavirus/immunocompromised/neutr_assay.txt")
neutr <- data.frame(neutr)
colnames(neutr) <- c("pat", "value", "point")
neutr$value <- as.numeric(neutr$value)
comp1 <- wilcox.test(neutr$value[neutr$point == "ref"], neutr$value[neutr$point == "aug"], paired = TRUE)
comp2 <- wilcox.test(neutr$value[neutr$point == "ref"], neutr$value[neutr$point == "jan"], paired = TRUE)
comp3 <- wilcox.test(neutr$value[neutr$point == "aug"], neutr$value[neutr$point == "jan"], paired = TRUE)
neutr$point <- factor(neutr$point, levels = c("ref", "aug", "jan"), 
                         labels = c("ref" = "Reference", "aug" = "Patient S\nAugust 20, 2020", 
                                    "jan" = "Patient S\nFebruary 19, 2021"))


p_neutr <-  ggplot(neutr, aes(x = as.factor(point), y = as.numeric(value))) +
  geom_segment(aes(x = 1.1, xend = 1.9, y = 300, yend = 300), size = 0.4,  color = "grey20") +
  geom_segment(aes(x = 2.1, xend = 2.9, y = 300, yend = 300), size = 0.4,  color = "grey20") +
  geom_segment(aes(x = 1.1, xend = 2.9, y = 420, yend = 420), size = 0.4,  color = "grey20") +
   annotate("text", x = 1.5, y=350, label = "0.009", size = 6) +
   annotate("text", x = 2.5, y=350, label = "0.01", size = 6) +
   annotate("text", x = 2, y=480, label = "0.593", size = 6) +
  geom_boxplot(outlier.colour = NA) +
  #geom_quasirandom(shape = 21, color = "black", alpha = 0.7, size = 3, fill = "lightsteelblue") +
  #geom_line(aes(group = pat), size = 0.5, alpha = 0.3, color = "grey50") +
  geom_dotplot(binaxis='y', 
               stackdir='center', 
               dotsize = 1.5, 
               fill="lightsteelblue", alpha = 0.8) +
  scale_x_discrete(name = "") +
  scale_y_log10(name = "Neutralising antibody titer", limits = c(8,  500), 
                breaks = trans_breaks("log2", function(x) 2^x),) +
  #stat_compare_means("wilcoxon", size = 3.5, paired = TRUE) +
  theme_bw() +
                    theme(legend.position = "none", 
                   legend.background = element_rect(), 
                   legend.title = element_text (colour="black", size = 14, face = "plain"),
                   legend.text = element_text(colour="black", size = 14, face = "plain"),
                   plot.title = element_text(colour="black", size = 14, face = "bold"),
                   axis.title = element_text (colour="black", size = 14, face = "plain"),
                   axis.text.y = element_text(colour="grey20", size = 14, face = "plain"),
                   axis.text.x = element_text(colour="grey20", size = 14, face = "plain", angle = 0, vjust = 0.5),
                   axis.line = element_line(colour="grey20", size = 0.6),
                   panel.background = element_rect(fill = "white", colour = "grey20"),
                   strip.text = element_text(size = 14))
p_neutr

pdf(file = "/home/jane/PhD_Skoltech/coronavirus/immunocompromised/plot_neutr.pdf",  
    width = 7, # The width of the plot in inches
    height = 5)
p_neutr
dev.off()

```

Binomial test fpr spike protein
12 in S over 47

```{r}
n_mut <- read.table("/home/jane/PhD_Skoltech/coronavirus/immunocompromised/n_mut.txt")
n_mut <- data.frame(n_mut)
colnames(n_mut) <- c("gene", "all", "nonsyn")
n_mut[,2:3] <- sapply(n_mut[,2:3], as.numeric)
coord$n <- nchar(coord$seq)
#p_exp <- sum(n_mut$all)/sum(coord$n)
#p_exp <- sum(n_mut$nonsyn)/sum(coord$n)
p_exp <- 25/sum(coord$n)
df <- c()
for (g in coord$gene){
  #test <- binom.test(n_mut$all[n_mut$gene == g], coord$n[coord$gene == g], alternative = "two.sided", p = p_exp, 
  #                   conf.level = 0.95)
  test <- binom.test(n_mut$nonsyn[n_mut$gene == g], coord$n[coord$gene == g], alternative = "two.sided", p = p_exp, 
                     conf.level = 0.95)
  df <- rbind(df, c(g, test$p.value, test$estimate, test$conf.int))
}

df <- data.frame(df)
colnames(df) <- c("gene", "p", "est", "low", "up")
df[,2:5] <-sapply(df[,2:5], as.numeric)

df$gene <- factor(df$gene, levels = rev(as.character(n_mut$gene)))
  
p_binom <- ggplot(df, aes(x = gene, y = est, color = p < 0.05)) +
  geom_errorbar(aes(ymin = low, ymax = up), size = 2) +
  geom_hline(yintercept = p_exp, linetype="dashed", color = "black", size = 1) +
  scale_color_manual( name = "P value", values = c("FALSE" = "grey40", "TRUE" = "red"),
                      labels = c("FALSE" = "ns", "TRUE" = "p < 0.05")) +
  scale_x_discrete(name = "") +
  scale_y_log10(name = "Mutation probability", limits = c()) +
  
  theme_bw() +
                   theme(legend.position = "top", 
                   legend.background = element_rect(), 
                   legend.title = element_text (colour="black", size = 12, face = "plain"),
                   legend.text = element_text(colour="black", size = 12, face = "plain"),
                   plot.title = element_text(colour="black", size = 14, face = "bold"),
                   axis.title = element_text (colour="black", size = 14, face = "plain"),
                   axis.text.y = element_text(colour="grey20", size = 14, face = "plain"),
                   axis.text.x = element_text(colour="grey20", size = 14, face = "plain", angle = 0, vjust = 0.5),
                   axis.line = element_line(colour="grey20", size = 0.6),
                   panel.background = element_rect(fill = "white", colour = "grey20"),
                   strip.text = element_text(size = 12)) + coord_flip()
  p_binom

```


